mindsdb/api/executor/sql_query/steps/fetch_dataframe_partition.py

```diff
@@ -0,0 +1,216 @@
+            threads = step.params['threads']
+            if isinstance(threads, int):
+                thread_count = threads
+                use_threads = True
+            if threads is True:
+                use_threads = True
+            if threads is False:
+                # disable even with ml task queue
+                use_threads = False
+
+        if use_threads:
+            return self.fetch_threads(run_query, query, thread_count=thread_count)
+        else:
+            return self.fetch_iterate(run_query, query)
+
+    def fetch_iterate(self, run_query, query):
+        # process batches in circle
+
+        results = []
+        while True:
+
+            # fetch batch
+            query2 = run_query.get_partition_query(self.current_step_num, query)
+            df, columns_info = self.dn.query(
+                query=query2,
+                session=self.session
+            )
+
+            if df is None or len(df) == 0:
+                break
+
+            # executing of sub steps can modify dataframe columns, lets memorise max tracking value
+            max_track_value = run_query.get_max_track_value(df)
+            sub_data = self.exec_sub_steps(df)
+
+            results.append(sub_data)
+
+            run_query.set_progress(df, max_track_value)
+
+        return self.concat_results(results)
+
+    def concat_results(self, results):
+        df_list = []
+        for res in results:
+            df, col_names = res.to_df_cols()
+            if len(df) > 0:
+                df_list.append(df)
+
+        data = ResultSet()
+        if len(df_list) > 0:
+            data.from_df_cols(pd.concat(df_list), col_names)
+
+        return data
+
+    def exec_sub_steps(self, df):
+        input_data = ResultSet()
+
+        input_data.from_df(
+            df,
+            table_name=self.table_alias[1],
+            table_alias=self.table_alias[2],
+            database=self.table_alias[0]
+        )
+
+        # execute with modified previous results
+        steps_data2 = self.steps_data.copy()
+        steps_data2[self.current_step_num] = input_data
+
+        sub_data = None
+        for substep in self.substeps:
+            sub_data = self.sql_query.execute_step(substep, steps_data=steps_data2)
+            steps_data2[substep.step_num] = sub_data
+        return sub_data
+
+    def fetch_threads(self, run_query, query, thread_count=None):
+        # process batches in threads
+
+        # create communication queues
+        queue_in = queue.Queue()
+        queue_out = queue.Queue()
+        self.stop_event = threading.Event()
+
+        if thread_count is None:
+            thread_count = get_max_thread_count()
+
+        # 3 tasks per worker during 1 batch
+        partition_size = int(run_query.batch_size / thread_count / 3)
+        # min partition size
+        if partition_size < 10:
+            partition_size = 10
+
+        # create N workers pool
+        workers = []
+        for i in range(thread_count):
```

Comment:
will be good to wrap all the following with `try/except/finally: stop_event.set()`

Reply:
wrapped